### PR TITLE
Add Sprint

### PR DESCRIPTION
**Pull request by adQuid**
**Brief Description of What This PR Does**

  Adds a simple sprint option so that extra ATP production and storage can still help. Sprint doubles speed, and can be combined with engulf mode (I've also made the penalty for engulfing more generous, so that timing engulfs to the last second isn't such a big part of hunting), but doubles movement cost. In addition, this adds a stress value that slowly increases while sprinting, upping the ATP cost so that no microbe can sprint forever.

Once merged new issues need to be opened for:
- Explaining sprint in the options menu
- Adding a sprint tutorial
- Adding a toolbar button for sprint that can indicate the current tiredness level

**Progress Checklist**

Note: before starting this checklist the issue should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
Is this ready for review yet or? Or should have this been a draft?

**Comment by adQuid** [comment]
> Is this ready for review yet or? Or should have this been a draft?

I did want people to look at it, but obviously it needs some work before it can merge. I don't know how you would categorize that.

File: src/microbe_stage/Microbe.cs
Comment (adQuid) [inline src/microbe_stage/Microbe.cs:1260]: I moved these lines so that the movement changes of engulfing and sprinting are applied after the effects of organelles, so that those bonuses are affected by states.

**Comment by adQuid** [comment]
@buckly90 would we be willing to merge this without the graphical indications of stress/sprinting status? If not, I'll add an impeded tag to denote that I'm blocked on UX work.

**Comment by hhyyrylainen** [comment]
I guess this could be merged even without the GUI and tutorial (otherwise players won't be able to even realize that this feature exists). Just need to fix the conflicts and pass code review and playtesting.

**Comment by RacerBG** [comment]
Additional text to the tutorial could be added directly in Weblate so it's not a deal breaker to merge this PR without it.

**Comment by hhyyrylainen** [comment]
Where would it be added? All the text boxes are pretty well full in English. This would need a separate tutorial box, and the timing of it should be considered (ie. *when* is it a good time to show the sprint tutorial).

**Comment by RacerBG** [comment]
> 
> 
> Where would it be added? All the text boxes are pretty well full in English. This would need a separate tutorial box, and the timing of it should be considered (ie. _when_ is it a good time to show the sprint tutorial).

MICROBE_STAGE_HELP_MESSAGE_1 seems like a good place but it depends from how much information you want to put in.

**Comment by hhyyrylainen** [comment]
No one reads the help menu to find new features of the game. It *has* to be a tutorial popup that appears during gameplay telling them about it. No one used the help menu to read about the controls before the tutorial about the movement buttons was added.

**Comment by adQuid** [comment]
I know @buckly90 was trying an alternative approach with his hotbar, consisting of putting interesting shapes on the screen and hoping players decide to hover over them for tooltips.

**Comment by hhyyrylainen** [comment]
That would help but for the future (for example controller players) I think this is too big of a feature (as in it makes it much easier to escape predators) to not have an explicit tutorial requiring the player to use it.

File: src/microbe_stage/MicrobeAI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeAI.cs:340]: Should the AI stop sprinting in certain cases?

File: src/microbe_stage/Microbe.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.cs:1388]: I think these should be represented by a property instead of having explicit setters like this. Similarly to how the microbe state transitions have extra logic implemented as a property.

File: src/microbe_stage/Microbe.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.cs:2003]: I think this should be rewritten to try to take the compound amount first and then check if enough could be taken, otherwise disable sprint.

File: src/microbe_stage/Microbe.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.cs:2127]: Unnecessary parentheses.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Besides the conflicts with master, a few code style comments, nothing too major the overall approach seems fine to me. Though, which key did you decide to use for sprinting? I think PC players would expect shift to be sprint. Also I think most often sprint is not a toggle. Would be excellent to have an option to have sprint as a toggle or a hold down key to sprint in the options menu.

**Comment by adQuid** [comment]
> Though, which key did you decide to use for sprinting? I think PC players would expect shift to be sprint. Also I think most often sprint is not a toggle. Would be excellent to have an option to have sprint as a toggle or a hold down key to sprint in the options menu.

It's set to shift, although it seems that that doesn't show in the code changes here. I deleted the local branch and re-pulled just to make sure it still works though. I'd also rather have it as a toggle, but I might need some help on the input handling.

File: src/microbe_stage/MicrobeAI.cs
Comment (adQuid) [reply src/microbe_stage/MicrobeAI.cs:340]: In the first iteration, I would suggest not. Sprint never lasts that long anyway, and an AI slowing down too soon and getting eaten will be more noticeable than an AI wasting a little ATP running away from something for a little too long.

**Comment by hhyyrylainen** [comment]
> but I might need some help on the input handling.

I would actually have to code this myself to provide exact info, but I think the following might work:
- Create 3 methods
- One is the same as currently reacting to toggling the key
- The two other ones are setup so that one reacts to key down and one reacts to key up
- Then hook up those 3 methods to react to a new option in the options class specifying which of them is active, and make the active methods control the sprint variable

Might need minor tweaks to the input system if the key up / down stuff doesn't work right off the bat. Might also need to configure one of the 2 approaches to not consume the input by returning false from the method.

File: src/microbe_stage/MicrobeAI.cs
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeAI.cs:340]: I guess I can open an issue about making the AI decide to start / stop sprinting in a more smart way...

File: simulation_parameters/Constants.cs
Comment (FuJa0815) [inline simulation_parameters/Constants.cs:251]:
```suggestion
    public const float MIN_ATP_PERCENT_TO_START_SPRINT = 0.9f;
```

File: simulation_parameters/Constants.cs
Comment (FuJa0815) [inline simulation_parameters/Constants.cs:256]:
```suggestion
    public const float MIN_ATP_PERCENT_TO_CONTINUE_SPRINT = 0.2f;
```

File: src/microbe_stage/Microbe.cs
Comment (FuJa0815) [inline src/microbe_stage/Microbe.cs:2035]: wouldn't it be better to have a seperate `SPRINT_STRESS_RELIEF_RATE`?

**Review by FuJa0815 — CHANGES_REQUESTED** [review]

File: src/microbe_stage/Microbe.cs
Comment (adQuid) [reply src/microbe_stage/Microbe.cs:2035]: My idea was that this was a universal stress increase rate for anything strenuous happening. Even if we don't want that to be the case in the future, I would argue it's still better to split the rates when the time comes, to avoid implying that there is some other non-sprint thing in the code that has this effect. Small potatoes though.

**Comment by adQuid** [comment]
Dropping this PR due to continuing issues in working with Godot UI and the babel scripts. The core of this code might come back once I've moved off my fork repo.

**Pull request closed without merging**
